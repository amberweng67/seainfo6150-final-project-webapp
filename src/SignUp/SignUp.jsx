import React, { useState } from 'react'
import styles from './SignUp.module.css';
import { useHistory } from "react-router-dom";

function getRequiredErrorMessage(type) {
    return `${type} is required`;
}

function getFormatErrorMessage(type) {
    return `${type} format is not correct`;
}

function getEqualErrorMessage(type1, type2) {
    return `${type1} and ${type2} do not have the same value`;
}

const CAREER_FIELDS = [
    'Product Manager',
    'UX Designer',
    'Software Engineer',
    'Data Scientist',
    'Business Analyst',
    'Other',
];

const EDUCATION_LEVELS = [
    'Bacholar',
    'Master',
    'Doctor',
    'MBA',
    'JD',
    'Other',
];

const SignUp = () => {
    const [email, setEmail] = useState('');
    const [password, setPassword] = useState('');
    const [repeatPassword, setRepeatPassword] = useState('');
    const [errorMessages, setErrorMessages] = useState([]);
    const [userName, setUserName] = useState('');
    const [phoneNumber, setPhoneNumber] = useState('');
    const [rememberMe, setRememberMe] = useState(true);
    const [sex, setSex] = useState('');
    const [careerField, setCareerField] = useState('');
    const [educationLevel, setEducationLevel] = useState('');

    const history = useHistory();
    const validateForm = () => {
        const errorMessages = [];
        const passwordFormValue = { type: 'password', label: 'Passowrd', value: password, required: true };
        const formValues = [
            {
                type: 'email', label: 'Email', value: email, required: true,
                validator: /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/,
            },
            {
                type: 'userName', label: 'User name', value: userName, required: true,
            },
            passwordFormValue,
            { type: 'repeatPassword', label: 'Repeat Password', value: repeatPassword, required: true, equalFormValue: passwordFormValue },
            {
                type: 'sex', label: 'Sex', value: sex, required: true,
            },
            {
                type: 'phoneNumber', label: 'Phone number', value: phoneNumber, required: false,
                validator: /[0-9]{10}/
            },
            {
                type: 'careerField', label: 'Career field', value: careerField, required: false,
            },
            {
                type: 'educationLevel', label: 'Education level', value: educationLevel,
            },
        ];
        for (const formValue of formValues) {
            if (!formValue.value && formValue.required) {
                errorMessages.push(getRequiredErrorMessage(formValue.label));
            } else if (formValue.validator) {
                if (formValue.value && !formValue.validator.test(formValue.value)) {
                    errorMessages.push(getFormatErrorMessage(formValue.label));
                }
            } else if (formValue.equalFormValue) {
                if (formValue.value !== formValue.equalFormValue.value) {
                    errorMessages.push(getEqualErrorMessage(formValue.label, formValue.equalFormValue.label));
                }
            }
        }
        return errorMessages;
    };

    const signUp = () => {
        const errorMessages = validateForm();
        if (errorMessages.length > 0) {
            setErrorMessages(errorMessages);
        } else {
            history.push('/login');
        }
    };

    const cancel = () => {
        history.push('/');
    }

    const selectSex = ({ target: { value } }) => {
        setSex(value);
    }

    return (
        <form className={styles.formContainer}>
            <div className={styles.container}>
                <h1>Sign Up</h1>
                <p>Please fill in this form to create an account.</p>
                <hr />

                <label htmlFor="email"><b>Email*</b></label>
                <input id="email" type="text" placeholder="Enter Email" name="email" required value={email} onChange={({ target: { value } }) => { setEmail(value) }} />

                <label htmlFor="user-name"><b>User Name*</b></label>
                <input id="user-name" type="text" placeholder="User name" name="user-name" value={userName} onChange={({ target: { value } }) => { setUserName(value) }} />

                <label htmlFor="psw"><b>Password*</b></label>
                <input id="psw" type="password" placeholder="Enter Password" name="psw" required value={password} onChange={({ target: { value } }) => { setPassword(value) }} />

                <label htmlFor="psw-repeat"><b>Repeat Password*</b></label>
                <input id="psw-repeat" type="password" placeholder="Repeat Password" name="psw-repeat" required value={repeatPassword} onChange={({ target: { value } }) => { setRepeatPassword(value) }} />

                <div className={styles.inputField}>
                    <label><b>Sex*</b></label>
                    <div>
                        <div>
                            <input type="radio" id="male" name="sex" value="male" onChange={selectSex} checked={sex === 'male'} />
                            <label htmlFor="male">Male</label>
                        </div>

                        <div>
                            <input type="radio" id="female" name="sex" value="female" onChange={selectSex} checked={sex === 'female'} />
                            <label htmlFor="female">Female</label>
                        </div>

                        <div>
                            <input type="radio" id="other" name="sex" value="other" onChange={selectSex} checked={sex === 'other'} />
                            <label htmlFor="other">Other</label>
                        </div>
                    </div>
                </div>

                <label htmlFor="phone-number"><b>Phone Number</b></label>
                <input id="phone-number" type="text" placeholder="Phone Number: 1234567890" name="phone-number" value={phoneNumber} onChange={({ target: { value } }) => { setPhoneNumber(value) }} />

                <label htmlFor="careerField"><b>Career Field</b></label>
                <div className={styles.inputField}>
                    <select name="careerField" id="careerField" onChange={(value) => {setCareerField(value)}}>
                        {CAREER_FIELDS.map(field=> <option key={field} value={field}>{field}</option>)}
                    </select>
                </div>

                <label htmlFor="educationlevel"><b>Education Level</b></label>
                <div className={styles.inputField}>
                    <select name="educationlevel" id="educationlevel" onChange={(value) => {setEducationLevel(value)}}>
                        {EDUCATION_LEVELS.map(level=> <option key={level} value={level}>{level}</option>)}
                    </select>
                </div>

                <label>
                    <input type="checkbox" checked={rememberMe} name="remember" onChange={() => {
                        setRememberMe(!rememberMe);
                    }} /> Remember me
                </label>

                <p>By creating an account you agree to our Terms & Privacy.</p>

                {errorMessages.length > 0 && <div className={styles.error}>
                    {errorMessages.map(msg => <div key={msg}>{msg}</div>)}
                </div>}
                <div className={styles.clearfix}>
                    <button className={styles.regularbutton} type="button" onClick={cancel}>Cancel</button>
                    <button className={styles.regularbutton} type="button" onClick={signUp}>Sign Up</button>
                </div>
            </div>
        </form>
    )
}

export default SignUp
