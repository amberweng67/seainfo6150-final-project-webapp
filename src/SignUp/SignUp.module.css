.container {
  padding: 16px;
  margin: 16px;
}

input[type=text]:focus, input[type=password]:focus {
  background-color: #ddd;
  outline: none;
}

hr {
  border: 1px solid #f1f1f1;
  margin-bottom: 25px;
}


.clearfix::after {
  content: "";
  clear: both;
  display: table;
}

.regularbutton{
  background-color: #001a3b;
  color: white;
  padding: 14px 20px;
  margin: 8px 0;
  border: none;
  cursor: pointer;
  width: 200px;
  border-radius: 20px;
  }

.error {
  color: red;
}

.inputField {
  margin: 8px 0;
}

@media screen and (max-width: 600px) {
  .formContainer  {
    width: 95%;
  }
}

@media screen and (max-width: 960px) and (min-width:601px) {
  .formContainer  {
    width: 75%;
  }
}