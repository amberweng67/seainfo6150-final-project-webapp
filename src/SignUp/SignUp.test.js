import React from "react";
import SignUp from "./SignUp";

describe("SignUp tests", () => {
  it("renders correctly", () => {
    const { container } = render(<SignUp />);
    expect(container).toMatchSnapshot();
  });
});
