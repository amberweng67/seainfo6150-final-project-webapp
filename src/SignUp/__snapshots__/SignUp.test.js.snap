// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUp tests renders correctly 1`] = `
<div>
  <form>
    <div
      class="container"
    >
      <h1>
        Sign Up
      </h1>
      <p>
        Please fill in this form to create an account.
      </p>
      <hr />
      <label
        for="email"
      >
        <b>
          Email*
        </b>
      </label>
      <input
        name="email"
        placeholder="Enter Email"
        required=""
        type="text"
        value=""
      />
      <label
        for="user-name"
      >
        <b>
          User Name*
        </b>
      </label>
      <input
        name="user-name"
        placeholder="User name"
        type="text"
        value=""
      />
      <label
        for="psw"
      >
        <b>
          Password*
        </b>
      </label>
      <input
        name="psw"
        placeholder="Enter Password"
        required=""
        type="password"
        value=""
      />
      <label
        for="psw-repeat"
      >
        <b>
          Repeat Password*
        </b>
      </label>
      <input
        name="psw-repeat"
        placeholder="Repeat Password"
        required=""
        type="password"
        value=""
      />
      <div
        class="inputField"
      >
        <label>
          <b>
            Sex*
          </b>
        </label>
        <div>
          <div>
            <input
              id="male"
              name="sex"
              type="radio"
              value="male"
            />
            <label
              for="male"
            >
              Male
            </label>
          </div>
          <div>
            <input
              id="female"
              name="sex"
              type="radio"
              value="female"
            />
            <label
              for="female"
            >
              Female
            </label>
          </div>
          <div>
            <input
              id="other"
              name="sex"
              type="radio"
              value="other"
            />
            <label
              for="other"
            >
              Other
            </label>
          </div>
        </div>
      </div>
      <label
        for="phone-number"
      >
        <b>
          Phone Number
        </b>
      </label>
      <input
        name="phone-number"
        placeholder="Phone Number: 1234567890"
        type="text"
        value=""
      />
      <label
        for="careerField"
      >
        <b>
          Career Field
        </b>
      </label>
      <div
        class="inputField"
      >
        <select
          id="careerField"
          name="careerField"
        >
          <option
            value="Product Manager"
          >
            Product Manager
          </option>
          <option
            value="UX Designer"
          >
            UX Designer
          </option>
          <option
            value="Software Engineer"
          >
            Software Engineer
          </option>
          <option
            value="Data Scientist"
          >
            Data Scientist
          </option>
          <option
            value="Business Analyst"
          >
            Business Analyst
          </option>
          <option
            value="Other"
          >
            Other
          </option>
        </select>
      </div>
      <label
        for="educationlevel"
      >
        <b>
          Education Level
        </b>
      </label>
      <div
        class="inputField"
      >
        <select
          id="educationlevel"
          name="educationlevel"
        >
          <option
            value="Bacholar"
          >
            Bacholar
          </option>
          <option
            value="Master"
          >
            Master
          </option>
          <option
            value="Doctor"
          >
            Doctor
          </option>
          <option
            value="MBA"
          >
            MBA
          </option>
          <option
            value="JD"
          >
            JD
          </option>
          <option
            value="Other"
          >
            Other
          </option>
        </select>
      </div>
      <label>
        <input
          checked=""
          name="remember"
          type="checkbox"
        />
         Remember me
      </label>
      <p>
        By creating an account you agree to our Terms & Privacy.
      </p>
      <div
        class="clearfix"
      >
        <button
          class="regularbutton"
          type="button"
        >
          Cancel
        </button>
        <button
          class="regularbutton"
          type="button"
        >
          Sign Up
        </button>
      </div>
    </div>
  </form>
</div>
`;
